tensorflow/core/framework/memory_types.cc

```diff
@@ -104,37 +108,58 @@ Status MemoryTypesForNode(const OpRegistryInterface* op_registry,
     return it != ndef.attr().end() && it->second.b();
   }();
 
-  bool has_kernel_def = status.ok() && !IsFunctionCallOp(ndef.op());
+  bool is_fn = IsFunctionCallOp(ndef.op());
+  bool has_kernel_def = status.ok() && !is_fn;
   auto host_memory_required = [&](const DataType& dt) {
     bool int32_on_device =
         has_kernel_def || device_type.type_string() == "TPU" || has_xla_compile;
     return DataTypeAlwaysOnHost(dt) || (dt == DT_INT32 && !int32_on_device);
   };
 
-  if (has_kernel_def) {
+  //  Edge cases:
+  //  1. If[Tcond=DT_BOOL, Tin=[DT_FLOAT,DT_INT32], Tout=[DT_FLOAT,DT_INT32]]
+  //     * Tcond marked HostMemory by kernel_def
+  //     * Tin[1] should be on host since it's INT32
+  //  2. ResourceGather[Tindices=DT_INT32](resource, indices)
+  //     * 'resource' marked HostMemory on the host
+  //     *  'indices' must be on the device (because the kernel expects it so)
+  //       - therefore, ignore the type
+
+  // Set all datatype to DEVICE_MEMORY by default, later on change it to
+  // HOST_MEMORY where it is required by the datatype.
+
+  if (has_kernel_def || is_fn) {
     // Gets the input/output names and their corresponding endpoint ranges.
     NameRangeMap inp_names;
     NameRangeMap out_names;
     TF_RETURN_IF_ERROR(
         NameRangesForNode(ndef, *op_def, &inp_names, &out_names));
 
     // Now that we know the size, fill with the default 'DEVICE_MEMORY'.
-    inp_mtypes->resize(GetTotal(inp_names), DEVICE_MEMORY);
-    out_mtypes->resize(GetTotal(out_names), DEVICE_MEMORY);
-
-    // Fills in host memory types based on the kernel def.
-    const auto& from_proto = kdef->host_memory_arg();
-    std::vector<string> host_memory_args(from_proto.begin(), from_proto.end());
-    MemoryTypesHelper(inp_names, &host_memory_args, inp_mtypes);
-    MemoryTypesHelper(out_names, &host_memory_args, out_mtypes);
-    if (!host_memory_args.empty()) {
-      return errors::InvalidArgument(
-          "HostMemory args '", absl::StrJoin(host_memory_args, "', '"),
-          "' not found in OpDef: ", SummarizeOpDef(*op_def));
+    // Stick to old behavior except for function calls
```

Comment:
People reading this comment after the PR is merged will likely have no idea what the old behavior is. Can you change the comment to describe the bhavior.